lib/active_merchant/billing/gateways/nuvei.rb

```diff
@@ -0,0 +1,298 @@
+        # The card details should be nested under cardData
+        card_details = post[:cardData].delete(:card)
+        post[:cardData].merge!(card_details)
+        commit(:general_credit, post)
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]').
+          gsub(%r(("cardNumber\\?":\\?")[^"\\]*)i, '\1[FILTERED]').
+          gsub(%r(("cardCvv\\?":\\?")\d+), '\1[FILTERED]').
+          gsub(%r(("cardHolderName\\?":\\?")\d+), '\1[FILTERED]').
+          gsub(%r(("merchantId\\?":\\?")\d+), '\1[FILTERED]').
+          gsub(%r(("merchantSiteId\\?":\\?")\d+), '\1[FILTERED]').
+          gsub(%r(("merchantKey\\?":\\?")\d+), '\1[FILTERED]')
+      end
+
+      private
+
+      def add_customer_ip(post, options)
+        return unless options[:ip_address]
+
+        post[:deviceDetails] = {
+          ipAddress: options[:ip_address]
+        }
+      end
+
+      def add_invoice(post, money, options)
+        post[:amount] = amount(money)
+        post[:currency] = (options[:currency] || currency(money))
+      end
+
+      def credit_card_hash(payment)
+        {
+          cardNumber: payment.number,
+          cardHolderName: payment.name,
+          expirationMonth: format(payment.month, :two_digits),
+          expirationYear: format(payment.year, :four_digits),
+          CVV: payment.verification_value
+        }
+      end
+
+      def add_payment(post, payment, key = :paymentOption)
+        if payment.is_a?(CreditCard)
+          post[key] = { card: credit_card_hash(payment) }
+        else
+          post[key] = { card: { cardToken: payment } }
+        end
+      end
+
+      def add_customer_names(full_name, payment_method)
+        split_names(full_name).tap do |names|
+          names[0] = payment_method&.first_name unless names[0].present? || payment_method.is_a?(String)
+          names[1] = payment_method&.last_name unless names[1].present? || payment_method.is_a?(String)
+        end
+      end
+
+      def add_address(post, payment, options, address_type = :billingAddress)
+        return unless address = options[:billing_address] || options[:address]
+
+        first_name, last_name = add_customer_names(address[:name], payment)
+
+        post[address_type] = {
+          email: options[:email],
+          country: address[:country],
+          phone: options[:phone] || address[:phone],
+          firstName: first_name,
+          lastName: last_name
+        }.compact
+      end
+
+      def current_timestamp
+        Time.now.utc.strftime('%Y%m%d%H%M%S')
+      end
+
+      def build_post_data(post, action)
+        post[:merchantId] = @options[:merchant_id]
+        post[:merchantSiteId] = @options[:merchant_site_id]
+        post[:timeStamp] = current_timestamp.to_i
+        post[:clientRequestId] = SecureRandom.uuid
+        post[:clientUniqueId] = SecureRandom.hex(16)
```

Comment:
probably, but not needed, the clientRequestId is only a API request identifier and the clientUniqueId is a unique transaction ID, since it is an optional param for the request, I think that is not needed use a GSF.